.docker/opt/configure-aiida.sh

```diff
@@ -33,6 +33,8 @@ if [[ ${NEED_SETUP_PROFILE} == true ]]; then
 
     # Setup and configure local computer.
     computer_name=localhost
+    cpu_count=`python -c 'import psutil; print(psutil.cpu_count(logical=False))'`
```

Comment:
should we add a check here that the returned value is a valid integer and otherwise unset the variable?

Reply:
Is `psutil` also guaranteed to be installed at this point? We currently have it in the `install_requires` of `aiida-core`, which I think answers that question in the positive, given that to run this script, AiiDA has to be installed. Might just pose a problem if we ever remove the last use of psutil in the actual module code of `aiida-core` and then it gets removed, because this use here is easily overlooked. Is there a specific reason to use `psutil` over `os.cpu_count()`?

Reply:
> should we add a check here that the returned value is a valid integer and otherwise unset the variable?

I believe the easiest way to protect against undefined behavior here would be to simply wrap the output like this:
```suggestion
    cpu_count=`python -c 'import psutil; print(int(psutil.cpu_count(logical=False)))'`
```
In this way the script will fail if the count cannot be determined.

> Is `psutil` also guaranteed to be installed at this point? We currently have it in the `install_requires` of `aiida-core`, which I think answers that question in the positive, given that to run this script, AiiDA has to be installed. Might just pose a problem if we ever remove the last use of psutil in the actual module code of `aiida-core` and then it gets removed, because this use here is easily overlooked. Is there a specific reason to use `psutil` over `os.cpu_count()`?

If `psutil` is ever removed as an aiida-core dependency, then the build of this image will simply fail, which would directly alert us to the problem. We could add a `pip install psutil~=5.6` command directly in this script if you prefer. We just need to make sure that this would not in any way create conflicts with the dependency specification of aiida-core.

We use `psutil` due to the reasons outlined here: https://github.com/aiidateam/aiida-core/issues/5117#issuecomment-913580903 .

Reply:
> In this way the script will fail if the count cannot be determined.

Are you sure? I think the exception will simply be written to `stderr`, the `cpu_count` variable will be empty, and the script will continue. Take this MWE:
```bash
#!/usr/bin/env bash
set -x
cpu_count=`python -c 'import psutil; print(non_existing.cpu_count(logical=False))'`
echo 'CPU_COUNT=' $cpu_count
```
Note that I replaced `psutil` with `non_existing` and so it will except but still get to the next line:
```
++ python -c 'import psutil; print(pstil.cpu_count(logical=False))'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
NameError: name 'pstil' is not defined
+ cpu_count=
+ echo CPU_COUNT=
CPU_COUNT=
```
You could set the `set -e` bash flag to automatically exit the script when a non-zero exit status is returned from any command.

> We could add a pip install psutil~=5.6 command directly in this script if you prefer.

Don't think that is a good idea. I think the script should simply fail if it cannot properly determine the value and not continue silently. Even better, if the env variable is defined (which would override it  anyway) maybe it is best to simply log the error as a warning and then continue anyway.

On a side-note: when is this script being run? When the image is built or when it is being run? That is to say, will this get the number of CPUs of the localhost that is building the image or the one that is running. I hope that it is the latter, because if not it would most likely still get the wrong value configured, would it not?

> We use psutil due to the reasons outlined here

Gotcha. Would have been useful to include this info in a comment, and/or at the very least in the commit message stating that you need `psutil`'s version because `os.cpu_count` does not give merely the physical CPUs, which is what is required here.

Reply:
> > In this way the script will fail if the count cannot be determined.
> 
> Are you sure? [...]

No, you are correct. I did not realize the `set -e` flag was not set.

> You could set the `set -e` bash flag to automatically exit the script when a non-zero exit status is returned from any command.

I opted for explicitly handling the failure of determination (which would only matter if the user did not override the variable) in https://github.com/aiidateam/aiida-core/pull/5161/commits/47f8a97b651fd2fd8e616ca356d46633185e2315 .
> 
> > We could add a pip install psutil~=5.6 command directly in this script if you prefer.
> 
> Don't think that is a good idea. I think the script should simply fail if it cannot properly determine the value and not continue silently. Even better, if the env variable is defined (which would override it anyway) maybe it is best to simply log the error as a warning and then continue anyway.
> 

This failure case would be handled by the error handling implemented in https://github.com/aiidateam/aiida-core/pull/5161/commits/47f8a97b651fd2fd8e616ca356d46633185e2315 .

> On a side-note: when is this script being run? When the image is built or when it is being run? That is to say, will this get the number of CPUs of the localhost that is building the image or the one that is running. I hope that it is the latter, because if not it would most likely still get the wrong value configured, would it not?
> 

Good catch, this script only makes sense to run on the host machine which is indeed the case.

> > We use psutil due to the reasons outlined here
> 
> Gotcha. Would have been useful to include this info in a comment, and/or at the very least in the commit message stating that you need `psutil`'s version because `os.cpu_count` does not give merely the physical CPUs, which is what is required here.

I will add a comment.